block manager specs (#1202)
rollkit/rollkit: block/block.md

# Block

## Abstract

> The block manager is a key component of full nodes and is responsible for block production or block syncing depending on the node type. Block syncing in this context includes retrieving the published blocks from the network (p2p network or DA network), validating them to raise fraud proofs upon validation failure, updating the state, and storing the validated blocks. A full node invokes multiple block manager functionalities in parallel, such as:

Reviewer: We should name the spec Block Manager if that is what we are describing.

Author: whats the right name? i think block is still valid as it will include various services like block management, block sync, header sync, etc?

Reviewer: It's just that you never talk about the block, only ever the services.  Plus you say that management includes syncing. Currently header sync is separate from block sync, but that feels like an implementation detail and it is reasonable for the block manager to be responsible for all parts of the block, and header is part of the block. The block being a struct/type, that the block manager handles.

Author: may be we can defer this for later?

Reviewer 2: I think Sevey just means to change block.md to block-manager.md

In `normal` mode, the block manager runs a timer, which is set to the `BlockTime` configuration parameter, and continuously produces blocks at `BlockTime` intervals.

In `lazy` mode, the block manager starts building a block when any transaction becomes available in the mempool. After the first notification of the transaction availability, the manager will wait for a 1 second timer to finish, in order to collect as many transactions from the mempool as possible. The 1 second delay is chosen in accordance with the default block time of 1s. The block manager also notifies the full node after every lazy block building.

#### Building the Block

The block manager of the sequencer nodes performs the following steps to produce a block:

* Call `CreateBlock` using executor
* Sign the block using `signing key` to generate commitment
* Call `ApplyBlock` using executor to generate an updated state
* Save the block, validators, and updated state to local store
* Add the newly generated block to `pendingBlocks` queue
* Publish the newly generated block to channels to notify other components of the sequencer node (such as block and header gossip)

### Block Publication to DA Network

> The block manager of the sequencer full nodes regularly publishes the produced blocks (that are pending in the `pendingBlocks` queue) to the DA network using the `DABlockTime` configuration parameter defined in the block manager config. In the event of failure to publish the block to the DA network, the manager will perform `maxSubmitAttempts` (default is 30) attempts and an exponential backoff interval between the attempts. The exponential backoff interval starts off at 100 milliseconds and it doubles in the next attempt and capped at `DABlockTime`. A successful publish event leads to the emptying of `pendingBlocks` queue and a failure event leads to proper error reporting and without emptying of `pendingBlocks` queue.

Reviewer: RE: `maxSubmitAttempts` we should just reference the variable and it's function. Adding the values in the specs adds another thing to track when things are updated. It is fine to use an example, like in the block times for rollup and DA layer.

Instead of listing the values of the constants/variables in the specs, we should reference where in the code the defaults are defined so that they can be referenced easily. 

This comment applies to all variables in the spec.

# Block

## Abstract

The block manager is a key component of full nodes (including sequencer) that is responsible for block production (in case of sequencer full node) or block syncing (syncing in this context includes retrieving the published blocks from the network, validating them to raise fraud proofs upon validation failure, and updating the state and storing the validated blocks). A full node invokes multiple of the block manager functionalities in parallel, such as:

* block production (in case of sequencer full nodes)
* block publication to DA network
* block retrieval from DA network
* block retrieval from Blockstore (which retrieves blocks from the p2p network)
* block syncing

## Protocol/Component Description

The block manager is initialized using several parameters as defined below:

> * `signing key`: used for signing the blocks after production

Author: its unique for sequencers and used for only signing the blocks that they produce.

* `config`: block manager configurations

> * `genesis`: initialize the block manager with genesis state

Author: done.

> * `localstore`: store blocks and states

Reviewer: is this a relative path? absolute path?

Author: is not a path. its a db. added additional information.

* `(mempool, proxyapp, eventbus)`: for initializing the executor (state transition function). mempool is also used in the manager to check for availability of transactions for lazy block production

> * `dalc`: to submit and retrieve blocks to DA network  

Reviewer: unless an abbreviation has previously been defined in the file, we should call out the full name in the description,
```suggestion
* `dalc`: the data availability light client used to submit and retrieve blocks to DA network  
```

* `blockstore`: to retrieve blocks from block store (p2p blocksync)

Block manager configuration options:

* `BlockTime`: time interval used for block production and block retrieval from Blockstore (default: 1 second)
* `DABlockTime`: time interval used for both block publication to DA network and block retrieval from DA network (default: 15 seconds)
* `DAStartHeight`: block retrieval from DA network starts from this height
* `NamespaceID`: 8 `byte` unique identifier of the rollup

### Block Production

> The sequencer (aka aggregator) type of full nodes run the block production logic. The sequencer can produce blocks under two modes: 1) normal and 2) lazy, which can be specified in the node configurations. In the normal mode, the sequencer runs a timer, which is set to the `BlockTime` configuration parameter of the block manager, and continuously publishes blocks at `BlockTime` intervals. Whereas, in the lazy mode, the manager starts building a block when any transaction becomes available in the mempool. The manager will wait for 1 second timer to finish as part of the block building process, after the first notification of the transaction availability, to collect as many transactions from the mempool as possible. The block manager also notifies the full node after every lazy block building.

Reviewer: nit: for easier reading i would split this into 3 smaller paragraphs

For the normal and lazy modes, are they defined with just the strings `normal` or `lazy`? or is there a different way? I think we should call that out here. 

Also this section is using 3 different names, sequencer, aggregator, and manager. We should probably define all terms at the top and when they will be used. Ideally we are consistent with language in the spec and throughout the code base. Either we say sequencer everywhere, or aggregator, or block manager in sequencer mode. I'm a little partial to the 3rd since this is the spec for the block manager and a sequencer is a node type that impacts how the block manager will perform. So writing the entire spec from the lens of a block manager always, and appropriately referencing the node type might be clearer imo. 

- [ ] The block production type is not called out in the previous config options

```suggestion
When the full node is operating as a sequencer (aka aggregator), the block manager runs the block production logic. There are two modes of block production, which can be specified in the block manager configurations: `normal` and `lazy`. 

In `normal` mode, the block manager runs a timer, which is set to the `BlockTime` configuration parameter, and continuously publishes blocks at `BlockTime` intervals. 

In `lazy` mode, the block manager starts building a block when any transaction becomes available in the mempool. After the first notification of the transaction availability, the manager will wait for a 1 second timer to finish before finalizing the block, in order to collect as many transactions from the mempool as possible. The block manager also notifies the full node after every lazy block building.
```

# Block

## Abstract

> The block manager is a key component of full nodes that is responsible for block production or block syncing depending on the node type. Block syncing in this context includes retrieving the published blocks from the network, validating them to raise fraud proofs upon validation failure, and updating the state and storing the validated blocks. A full node invokes multiple of the block manager functionalities in parallel, such as:

Reviewer 3: ```suggestion
The block manager is a key component of full nodes and is responsible for block production or block syncing depending on the node type. Block syncing in this context includes retrieving the published blocks from the network, validating them to raise fraud proofs upon validation failure, updating the state, and storing the validated blocks. A full node invokes multiple block manager functionalities in parallel, such as:
```

> * block production (only for sequencer full nodes)

Author: i am using sequencer throughout.

> * block production (only for sequencer full nodes)

Reviewer 3: in the code we use aggregator everywhere, maybe we use aggregator here or change it to sequencer everywhere in the code later? can be done in a follow-up. Related issue: https://github.com/rollkit/rollkit/issues/796

> * block production (only for sequencer full nodes)

Author: as part of #796, we can fix the specs too? i am not bought on either :P

> * block production (only for sequencer full nodes)

Reviewer 2: If we have an internal audit the naming convention should be the same. So I would say 796 should be part of VA and the specs work. We can leave it as a sequencer I think.

* block publication to DA network
* block retrieval from DA network
* block retrieval from Blockstore (which retrieves blocks from the p2p network)
* block syncing

> ## Protocol/Component Description

Reviewer: I think I'd take the approach of documenting the structs in the specs themselves first. 
If we then see the same things documented in multiple specs, that would help us see areas we could refactor the code as well.

The block manager is initialized using several parameters as defined below:

> * `signing key`: used for signing the blocks after production

Reviewer 3: ```suggestion
* `signing key`: used for signing a block after it is created
```

* `config`: block manager configurations (see config options below)
* `genesis`: initialize the block manager with genesis state

> * `localstore`: store blocks and states

Author: yes it is persistent as the default is disk storage when the db path is specified in the config.

> * `localstore`: store blocks and states

Author: added an assumption for that.

> * `localstore`: store blocks and states

Author: if you specify in the config, it will be persistent storage, otherwise in-memory only.

> * `localstore`: store blocks and states

Reviewer 2: I think the default should be persistent storage right? Otherwise, people will always have to start from scratch. Maybe a flag ephemeral for it to stay in memory. That's more of a devEx question. cc @jcstein @MSevey thoughts ?

> * `localstore`: store blocks and states

Reviewer 4: I think the default being persistent would make sense. for example, if I ran a Polaris rollup and had deployed smart contracts to it for my applications, I would expect that starting back up my local environment would persist the state of the EVM chain by default. wouldn't this cause issues though if the devnet were stopped and restarted?

it should still be possible to wipe the state, too, so that when a user decides they want to, they can wipe the old chain history and start fresh.

i'm curious to see what Sevey thinks here, too

> * `localstore`: store blocks and states

Reviewer: Sorry missed this ping. 

I agree that the default should be persistent storage. The reason being that all defaults should be production defaults. The expectation of a production node is that if it stops, it will restart with the previous data. 

Having a flag for testing/dev/in-memory etc then makes sense.

> * `localstore`: store blocks and states

Author: for cosmos-sdk apps, the default is always persistent storage (meaning, the db_dir is set to `data` by default) and rollkit will simply inherit that.

* `(mempool, proxyapp, eventbus)`: for initializing the executor (state transition function). mempool is also used in the manager to check for availability of transactions for lazy block production
* `dalc`: the data availability light client used to submit and retrieve blocks to DA network  

> * `blockstore`: to retrieve blocks from block store (p2p blocksync)

Reviewer 3: ```suggestion
* `blockstore`: to retrieve blocks gossiped over the p2p network
```

Block manager configuration options:

> * `BlockTime`: time interval used for block production and block retrieval from Blockstore (default: 1 second)

Reviewer 3: We should be consistent with how we use the word "block store" or "Blocktore" everywhere
```suggestion
* `BlockTime`: time interval used for block production and block retrieval from block store (default: 1 second)
```

* `DABlockTime`: time interval used for both block publication to DA network and block retrieval from DA network (default: 15 seconds)
* `DAStartHeight`: block retrieval from DA network starts from this height

> * `NamespaceID`: 8 `byte` unique identifier of the rollup

Reviewer 3: Related issue: https://github.com/rollkit/rollkit/issues/1062

### Block Production

When the full node is operating as a sequencer (aka aggregator), the block manager runs the block production logic. There are two modes of block production, which can be specified in the block manager configurations: `normal` and `lazy`.

> In `normal` mode, the block manager runs a timer, which is set to the `BlockTime` configuration parameter, and continuously publishes blocks at `BlockTime` intervals.

Author: done.

> In `lazy` mode, the block manager starts building a block when any transaction becomes available in the mempool. After the first notification of the transaction availability, the manager will wait for a 1 second timer to finish before finalizing the block, in order to collect as many transactions from the mempool as possible. The block manager also notifies the full node after every lazy block building.

Reviewer 3: Any reason why we wait for 1 second specifically? Maybe we should mention the reason here.

Author: the reason is "in order to collect as many transactions from the mempool as possible". specifically 1s to be inline with our default blocktime of 1s.

Author: added `The 1 second delay is chosen in accordance with the default block time of 1s.` in addition.

#### Building the Block

> The block manager of the sequencer nodes perform following steps to produce a block:

Reviewer 3: ```suggestion
The block manager of the sequencer nodes performs the following steps to produce a block:
```

* Call `CreateBlock` using executor

> * Sign the block to generate commitment

Reviewer 3: ```suggestion
* Sign the block using `signing key` to generate commitment
```

* Call `ApplyBlock` using executor to generate an updated state
* Save the block, validators, and updated state to local store
* Add the newly generated block to `pendingBlocks` queue
* Publish the newly generated block to channels to notify other components of the sequencer node (such as block and header gossip)

> ### Block Publication to DA Network

Author: for failure attempts, the queue has no effect, meaning we won't remove the blocks (or empty it). 

yes, the block production proceeds as usual even if the DA attempt keeps failing and the blocks keep on getting accumulated in the pending queue. Not sure if it impacts the transactions. 

By nonce clash, you mean DA nonce? the rollup nonce should not clash as the sequencer state is updated as and when the blocks are produced irrespective of whether they make it to DA or not.

> ### Block Publication to DA Network

Author: if i understand this correctly, the DA tx nonce for the first attempt is, let's say, 1 and the DA tx fails, which should not increment the nonce for the submitter account and next time when a new DA tx is made (which includes 15 blocks), the same nonce should get used, whats the issue here?

> The block manager of the sequencer full nodes regularly publish the produced blocks (that are pending in the `pendingBlocks` queue) to DA network using the `DABlockTime` configuration parameter defined in the block manager config. In the event of failure to publish the block to DA network, the manager will perform `maxSubmitAttempts` (default is 30) attempts and a exponential backoff interval between the attempts. The exponential backoff interval starts off at 100 milliseconds and it doubles in the next attempt and capped at `DABlockTime`. A successfull publish event leads to emptying of `pendingBlocks` queue and a failure event leads to proper error reporting.

Reviewer 3: ```suggestion
The block manager of the sequencer full nodes regularly publishes the produced blocks (that are pending in the `pendingBlocks` queue) to the DA network using the `DABlockTime` configuration parameter defined in the block manager config. In the event of failure to publish the block to the DA network, the manager will perform `maxSubmitAttempts` (default is 30) attempts and an exponential backoff interval between the attempts. The exponential backoff interval starts off at 100 milliseconds and it doubles in the next attempt and capped at `DABlockTime`. A successful publish event leads to the emptying of `pendingBlocks` queue and a failure event leads to proper error reporting.
```

> The block manager of the sequencer full nodes regularly publishes the produced blocks (that are pending in the `pendingBlocks` queue) to the DA network using the `DABlockTime` configuration parameter defined in the block manager config. In the event of failure to publish the block to the DA network, the manager will perform `maxSubmitAttempts` (default is 30) attempts and an exponential backoff interval between the attempts. The exponential backoff interval starts off at 100 milliseconds and it doubles in the next attempt and capped at `DABlockTime`. A successful publish event leads to the emptying of `pendingBlocks` queue and a failure event leads to proper error reporting.

Author: the blob size is already imposed by the blob module when we call `blob.NewBlobV0(c.namespace.Bytes(), data)`:
```
if len(data) == 0 || len(data) > appconsts.DefaultMaxBytes {
     return nil, fmt.Errorf("blob data must be > 0 && <= %d, but it was %d bytes", appconsts.DefaultMaxBytes, len(data))
}
```

> The block manager of the sequencer full nodes regularly publishes the produced blocks (that are pending in the `pendingBlocks` queue) to the DA network using the `DABlockTime` configuration parameter defined in the block manager config. In the event of failure to publish the block to the DA network, the manager will perform `maxSubmitAttempts` (default is 30) attempts and an exponential backoff interval between the attempts. The exponential backoff interval starts off at 100 milliseconds and it doubles in the next attempt and capped at `DABlockTime`. A successful publish event leads to the emptying of `pendingBlocks` queue and a failure event leads to proper error reporting.

Reviewer 2: Avail has a smaller max blob size than Celestia; how do we ensure we can support both?

> The block manager of the sequencer full nodes regularly publishes the produced blocks (that are pending in the `pendingBlocks` queue) to the DA network using the `DABlockTime` configuration parameter defined in the block manager config. In the event of failure to publish the block to the DA network, the manager will perform `maxSubmitAttempts` (default is 30) attempts and an exponential backoff interval between the attempts. The exponential backoff interval starts off at 100 milliseconds and it doubles in the next attempt and capped at `DABlockTime`. A successful publish event leads to the emptying of `pendingBlocks` queue and a failure event leads to proper error reporting.

Author: created an issue: https://github.com/rollkit/rollkit/issues/1233, i think the DA interface can provide a method for blob size limit?

### Block Retrieval from DA Network

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of error, a retry logic kicks in after a delay of 100 milliseconds delay between every retries and 10 retries, after which the error is logged and `daHeight` counter is not incremented, which basically results in the intentional stall of the block retrieval logic. In block `NotFound` scenario does not result in an error and it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first the hashes of the blocks which are successfully retrieved are marked as hard confirmed and are sent for application (or state update). The successful state update triggers a fresh DA and blockstore pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Reviewer 3: ```suggestion
The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.
```

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Reviewer 3: My understanding is that "applying a block" and a "state update" are the same thing and we should just pick one, maybe just get rid of the "(or state update)". We do not apply a block that we have already seen on the P2P layer, we keep track of blocks seen in the `blockCache`.

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Author: Also added an a clarification: `The block manager does not re-apply the block again (in other words, create a new updated state and persist it) when a block was initially applied using p2p block sync, but later was hard confirmed by DA retrieval. The block is only set hard confirmed in this case.`

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Author: we first signal and then apply the block. the signal here triggers next pull.

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Reviewer 3: I think `ApplyBlock` would also set the ISRs in the passed-in block in the future so maybe not ready only then? But yes it's read-only currently.

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Author: yes, the pull is for next DAHeight

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Author: state machine

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Author: currently we don't try to re-apply the block upon DA retrieval success. this is not needed because the block is previously applied upon p2p retrieval. note that, if the block is changed, it won't have the same block hash, hence no need for re-apply and simply marking it as hard confirmed should be sufficient?

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Reviewer 2: Do we only NOT apply the block if we get it over p2p? Then it is fine. IF we check the block hash before. Do we?  

My question rephrased: What if we don't get it over p2p but only from the DA-Layer? Would we hard confirm it even *before*  applying to the state?

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Author: yes, if the block is only coming from DA, it is applied. Basically, the logic does make sure to only apply the block once, thats all.

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Reviewer 2: Ok, I think what we should do here is define hard confirmed because we think about it probably differently.

> The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.

Reviewer 2: Follow up: https://github.com/rollkit/rollkit/issues/1240

#### About Soft/Hard Confirmations

> The block manager retrieves blocks from both p2p network and the underlying DA network, because the blocks are available in the p2p network faster and DA retreival is slower (e.g., 1 second vs 15 seconds). The blocks retrieved from the p2p network are only marked as soft confirmed until the DA retrieval succeeds on those blocks at which time they are marked hard confirmed. The hard confirmations can be considered as finality.

Reviewer 3: ```suggestion
The block manager retrieves blocks from both the p2p network and the underlying DA network because the blocks are available in the p2p network faster and DA retrieval is slower (e.g., 1 second vs 15 seconds). The blocks retrieved from the p2p network are only marked as soft confirmed until the DA retrieval succeeds on those blocks and they are marked hard confirmed. The hard confirmations can be considered to have a higher level of finality.
```

### Block Retrieval from BlockStore (P2P BlockSync)

> The block manager of the full nodes regularly pulls blocks from the blockstore (which in turn uses the p2p network for syncing the blocks) at `BlockTime` intervals and starts off with a blockstore height zero. Every time the blockstore height is higher than the last seen height, the newest blocks are pulled from the blockstore and sent for application (or state update), along with updating the last seen blockstore height.

Reviewer 3: ```suggestion
The block manager of the full nodes regularly pulls blocks from the block store (which in turn uses the p2p network for syncing the blocks) at `BlockTime` intervals and starts off with a block store height of zero. Every time the block store height is higher than the last seen height, the newest blocks are pulled from the block store and sent to be applied (or state update), along with updating the last seen block store height.
```

### Block Syncing

> The block manager stores and applies the block everytime a new block is retrieved either via blockstore or DA network. Block syncing involves:

Reviewer 3: ```suggestion
The block manager stores and applies the block every time a new block is retrieved either via the blockstore or DA network. Block syncing involves:
```

> The block manager of the full nodes regularly pulls blocks from the block store (which in turn uses the p2p network for syncing the blocks) at `BlockTime` intervals and starts off with a block store height of zero. Every time the block store height is higher than the last seen height, the newest blocks are pulled from the block store and sent to be applied (or state update), along with updating the last seen block store height.

Reviewer 3: Same comment as above, we should get rid of the "(or state update)".

> * `ApplyBlock` using executor: validate the block, executes the block (applies the transactions), captures the validator updates, and creates an update state.

Author: "updated state"

> * `ApplyBlock` using executor: validate the block, executes the block (applies the transactions), captures the validator updates, and creates an update state.

Author: it doesn't update the state, only creates an updated state which is later applied to actually update the state. ApplyBlock is read-only

* `Commit` using executor: commit the execution and changes, update mempool, and publish events
* Store the block, the validators, and the updated state.

## Message Structure/Communication Format

The communication between the block manager and executor:

* `InitChain`: using the genesis, a set of parameters, and validator set to invoke `InitChainSync` on the proxyApp to obtain initial `appHash` and initialize the state.
* `Commit`: commit the execution and changes, update mempool, and publish events.
* `CreateBlock`: prepare a block by polling transactions from mempool.
* `ApplyBlock`: validate the block, execute the block (apply transactions), validator updates, create and return updated state

The communication between the full node and block manager:

* Notify when the block is published
* Notify when the block is done lazy building

## Assumptions and Considerations

> * The block manager loads the initial state from the local store and uses genesis if not found in the local store, when the node (re)starts.

Reviewer 2: I would add what it means to have a successful submission to the da-layer and what it does not mean.

Author: i think you are talking about the tx validity rules. since this is performed in the mempool pkg, we should defer it to mempool specs. once the tx is validated and lands in the mempool, it is picked up by the block manager to compose a block, etc. the possible failure scenarios afterwards would be ApplyBlock and SubmitBlockToDA. The specs for ApplyBlock should be covered in the executor specs and SubmitBlockToDA scenario should be covered in the DA specs.

Reviewer 2: I was talking about a Celestia Transaction. A rollup Block i will edit my comment, cc @S1nus for mempool pkg @tuxcanfly for celestia, @Manav-Aggarwal for executor and @tzdybal for DA-Interface. Let's try figure this out.

Reviewer 2: cc @gupadhyaya I do believe Manager should be aware of those interactions as we might be making the wrong assumptions here

Author: created https://github.com/rollkit/rollkit/issues/1236

## Abstract

> The block manager is a key component of full nodes that is responsible for block production or block syncing depending on the node type. Block syncing in this context includes retrieving the published blocks from the network, validating them to raise fraud proofs upon validation failure, and updating the state and storing the validated blocks. A full node invokes multiple of the block manager functionalities in parallel, such as:

Author: p2p or DA network

> The block manager is a key component of full nodes that is responsible for block production or block syncing depending on the node type. Block syncing in this context includes retrieving the published blocks from the network, validating them to raise fraud proofs upon validation failure, and updating the state and storing the validated blocks. A full node invokes multiple of the block manager functionalities in parallel, such as:

Reviewer 3: Should we change "the network" to "the p2p network or DA layer" then?

> The block manager is a key component of full nodes that is responsible for block production or block syncing depending on the node type. Block syncing in this context includes retrieving the published blocks from the network, validating them to raise fraud proofs upon validation failure, and updating the state and storing the validated blocks. A full node invokes multiple of the block manager functionalities in parallel, such as:

Author: `the network (p2p network or DA network)`
